packages/ember-glimmer/lib/syntax/mount.js

```diff
@@ -113,25 +121,35 @@ class MountManager extends AbstractManager {
 
     engine.boot();
 
-    return engine;
+    return { engine, args };
   }
 
-  layoutFor(definition, engine, env) {
+  layoutFor(definition, { engine }, env) {
     let template = engine.lookup(`template:application`);
     return env.getCompiledBlock(OutletLayoutCompiler, template);
   }
 
-  getSelf(engine) {
+  getSelf(bucket) {
+    let { engine, args } = bucket;
+
     let applicationFactory = engine[FACTORY_FOR](`controller:application`);
-    let factory = applicationFactory || generateControllerFactory(engine, 'application');
-    return new RootReference(factory.create());
+    let controllerFactory = applicationFactory || generateControllerFactory(engine, 'application');
+    let controller = bucket.controller = controllerFactory.create();
```

Comment:
Pardon my lack of rendering knowledge, but is mutating the `bucket` input here kosher?